nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -265,6 +265,55 @@ func (w *deploymentWatcher) PromoteDeployment(
 	return nil
 }
 
+// autoPromoteDeployment creates a synthetic promotion request, and upserts it for processing
+func (w *deploymentWatcher) autoPromoteDeployment(allocs []*structs.AllocListStub) error {
+	d := w.getDeployment()
+	if !d.HasPlacedCanaries() || !d.RequiresPromotion() {
+		return nil
+	}
+
+	// Find the promotable groups
+	groups := []string{}
+
+TASKGROUP:
+	for tk, tv := range d.TaskGroups {
+		if !tv.AutoPromote || tv.DesiredCanaries != len(tv.PlacedCanaries) {
+			continue
```

Comment:
Mind if you elaborate on auto promotion and multiple task groups? I would have expected auto promotion of a job to take place if all its task group were set for AutoPromote or if AutoPromote was a job level property.

Reply:
auto_promote will only update the job if all the task groups are healthy and marked for auto_promote, it will otherwise stay in the pending state